File: react-app/src/scripts/skill.ts
// "Bleed", "Poison", "Burn" -> DoT
// "Freeze" -> DoT & CC
// "Stun", "Paralysis", "Sleep" -> CC
// "Silence", "Confusion", "Blindness", "Weaken" -> Debuff

export enum StatusEffectType {
    Burn = "Burn",
    Poison = "Poison",
    Bleed = "Bleed",
    Stun = "Stun",
    Freeze = "Freeze",
    Paralysis = "Paralysis",
    Sleep = "Sleep",
    Confusion = "Confusion",
    Blindness = "Blindness",
    Charm = "Charm",
    Fear = "Fear",
    Weaken = "Weaken"
}

interface StatusEffectData {
    type: StatusEffectType;
    duration: number;
    defaultValue: number; 
    coef: Record<string,number>;
    accuracy: number;
}

export interface FinalStatusEffectData {
    type: StatusEffectType;
    duration: number;
    value: number; 
    accuracy: number;
}

class StatusEffect {
    type: StatusEffectType;
    duration: number;
    defaultValue: number; 
    coef: Record<string,number>;
    accuracy: number;

    constructor(data: StatusEffectData){
        this.type = data.type;
        this.defaultValue = data.defaultValue;
        this.coef = data.coef;
        this.accuracy = data.accuracy;
        this.duration = data.duration;
    }

    doAttack(stats: Record<string, any>): FinalStatusEffectData {
        let totalValue = this.defaultValue;
        for (const [key, value] of Object.entries(this.coef)){
            totalValue += stats[key] * value;
        }
        return {
            type: this.type,
            duration: this.duration,
            value: totalValue, 
            accuracy: this.accuracy
        }
    }

    getTotalValue(stats: Record<string, any>): number {
        let totalValue = this.defaultValue;
        for (const [key, value] of Object.entries(this.coef)){
            totalValue += stats[key] * value;
        }
        return totalValue;
    }

    toDict(): StatusEffectData {
        return {
            type: this.type,
            duration: this.duration,
            defaultValue: this.defaultValue,
            coef: this.coef,
            accuracy: this.accuracy
        }
    }
}

export enum AttackType {
    melee = "melee",
    magic = 'magic'
}

export interface AttackData {
    name: string;
    type: AttackType;
    defaultDamage:number;
    coef: Record<string, number>;
    count: number;
    penetration: number;
    accuracy: number;
    cooldown: number;
    statusEffect: StatusEffectData | null;
    curCooldown: number;
}

export interface FinalAttackData {
    name: string;
    type: AttackType;
    damage: number;
    count: number;
    penetration: number;
    accuracy: number;
    statusEffect: FinalStatusEffectData | null;
}

export class Attack {
    name: string;
    type: AttackType;
    defaultDamage:number;
    coef: Record<string, number>;
    count: number;
    penetration: number;
    accuracy: number;
    cooldown: number;
    statusEffect: StatusEffect | null;
    
    private curCooldown: number;

    constructor(data: AttackData) {
        this.name = data.name;
        this.type = data.type;
        this.defaultDamage = data.defaultDamage;
        this.coef = data.coef;
        this.count = data.count;
        this.penetration = data.penetration;
        this.accuracy = data.accuracy;
        this.cooldown = data.cooldown;
        if (data.statusEffect)
            this.statusEffect = new StatusEffect(data.statusEffect);
        else
            this.statusEffect = null;
        this.curCooldown = 0;
    }

    doAttack(stats: Record<string, any>): FinalAttackData {
        let totalDamage = this.defaultDamage;

        for (const [key, value] of Object.entries(this.coef)){
            totalDamage += stats[key] * value;
        }

        this.curCooldown += this.cooldown;

        return {
            name: this.name,
            type: this.type,
            damage:  Math.floor(totalDamage),
            count: this.count,
            penetration: this.penetration,
            accuracy: this.accuracy,
            statusEffect: this.statusEffect ? this.statusEffect.doAttack(stats) : null,
        };
    }

    getTotalDamage(stats: Record<string, any>): number {
        let totalDamage = this.defaultDamage;

        for (const [key, value] of Object.entries(this.coef)){
            totalDamage += stats[key] * value;
        }
        return Math.floor(totalDamage);
    }

    isAvailable(): boolean {
        return this.curCooldown === 0;
    }

    reduceCooldown(value: number): void {
        this.curCooldown = this.curCooldown - value >= 0 ? this.curCooldown - value : 0;
    }

    toDict(): AttackData {
        return {
            name: this.name,
            type: this.type,
            defaultDamage: this.defaultDamage,
            coef: this.coef,
            count: this.count,
            penetration: this.penetration,
            accuracy: this.accuracy,
            cooldown: this.cooldown,
            statusEffect: this.statusEffect ? this.statusEffect.toDict(): null,
            curCooldown: this.curCooldown
        };
    }
}

export interface DefendData {
    name: string;
    type: string;
    defaultValue:number;
    coef: Record<string, number>;
    duration: number;
    cooldown: number;
    curCooldown: number;
}

export class Defend {
    name: string;
    type: string; // defense up, shield, dodge
    defaultValue:number;
    coef: Record<string, number>;
    duration: number;
    cooldown: number;
    
    private curCooldown: number;

    constructor(data: DefendData) {
        this.name = data.name;
        this.type = data.type;
        this.defaultValue = data.defaultValue;
        this.coef = data.coef;
        this.duration = data.duration;
        this.cooldown = data.cooldown;
        this.curCooldown = 0;
    }

    doDefend(stats: Record<string, any>): Record<string, any> {
        let totalValue = this.defaultValue;
        for (const [key, value] of Object.entries(this.coef)) {
            totalValue += stats[key] * value;
        }
        this.curCooldown += this.cooldown;
        return {
            name: this.name,
            type: this.type,
            value:  Math.floor(totalValue),
            duration: this.duration,
        };
    }

    isAvailable(): boolean {
        return this.curCooldown === 0;
    }

    reduceCooldown(value: number): void {
        this.curCooldown = this.curCooldown - value >= 0 ? this.curCooldown - value : 0;
    }

    getTotalValue(stats: Record<string, any>): number {
        let totalValue = this.defaultValue;
        for (const [key, value] of Object.entries(this.coef)) {
            totalValue += stats[key] * value;
        }
        return totalValue;
    }

    toDict(): DefendData {
        return {
            name: this.name,
            type: this.type,
            defaultValue: this.defaultValue,
            coef: this.coef,
            duration: this.duration,
            cooldown: this.cooldown,
            curCooldown: this.curCooldown
        };
    }
}

export interface SmiteData {
    name: string;
    type: string;
    defaultValue:number;
    coef: Record<string, number>;
    duration: number;
    cooldown: number;
    curCooldown: number;
}

export class Smite {
    name: string;
    type: string; // hp_scailing, damage, stun
    defaultValue:number;
    coef: Record<string, number>;
    duration: number;
    cooldown: number;
    
    private curCooldown: number;

    constructor(data: SmiteData) {
        this.name = data.name;
        this.type = data.type;
        this.defaultValue = data.defaultValue;
        this.coef = data.coef;
        this.duration = data.duration;
        this.cooldown = data.cooldown;
        this.curCooldown = 0;
    }

    doSmite(stats: Record<string, any>): Record<string, any> {
        let totalValue = this.defaultValue;
        for (const [key, value] of Object.entries(this.coef)) {
            totalValue += stats[key] * value;
        }
        this.curCooldown += this.cooldown;
        return {
            name: this.name,
            type: this.type,
            value:  Math.floor(totalValue),
            duration: this.duration,
        };
    }

    isAvailable(): boolean {
        return this.curCooldown === 0;
    }

    reduceCooldown(value: number): void {
        this.curCooldown = this.curCooldown - value >= 0 ? this.curCooldown - value : 0;
    }

    getTotalValue(stats: Record<string, any>): number {
        let totalValue = this.defaultValue;
        for (const [key, value] of Object.entries(this.coef)) {
            totalValue += stats[key] * value;
        }
        return totalValue;
    }
    
    toDict(): SmiteData {
        return {
            name: this.name,
            type: this.type,
            defaultValue: this.defaultValue,
            coef: this.coef,
            duration: this.duration,
            cooldown: this.cooldown,
            curCooldown: this.curCooldown
        };
    }
}

File: fastapi-app/skill.py
from typing import Dict, Any

# StatusEffect class
class StatusEffect:
    def __init__(self, effect_type: str, duration: int, default_value: int, coef: Dict[str, int], accuracy: int):
        self.type = effect_type
        self.duration = duration
        self.default_value = default_value
        self.coef = coef
        self.accuracy = accuracy

    def to_dict(self) -> Dict[str, Any]:
        return {
            "type": self.type,
            "duration": self.duration,
            "defaultValue": self.default_value,
            "coef": self.coef,
            "accuracy": self.accuracy
        }

# Attack class
class Attack:
    def __init__(self, name: str, attack_type: str, default_damage: int, coef: Dict[str, int], count: int, 
                 penetration: int, accuracy: int, cooldown: int, status_effect: StatusEffect):
        self.name = name
        self.type = attack_type
        self.default_damage = default_damage
        self.coef = coef
        self.count = count
        self.penetration = penetration
        self.accuracy = accuracy
        self.cooldown = cooldown
        self.status_effect = status_effect
        self.cur_cooldown = 0

    def do_attack(self, stats: Dict[str, Any]) -> Dict[str, Any]:
        total_damage = self.default_damage

        for key, value in self.coef.items():
            total_damage += stats.get(key, 0) * value

        self.cur_cooldown += self.cooldown

        return {
            "type": self.type,
            "damage": total_damage,
            "count": self.count,
            "penetration": self.penetration,
            "accuracy": self.accuracy,
            "statusEffect": self.status_effect.to_dict(),
        }

    def is_available(self) -> bool:
        return self.cur_cooldown == 0

    def reduce_cooldown(self, value: int) -> None:
        self.cur_cooldown = max(self.cur_cooldown - value, 0)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "type": self.type,
            "defaultDamage": self.default_damage,
            "coef": self.coef,
            "count": self.count,
            "penetration": self.penetration,
            "accuracy": self.accuracy,
            "cooldown": self.cooldown,
            "statusEffect": self.status_effect.to_dict(),
            "curCooldown": self.cur_cooldown
        }

# Defend class
class Defend:
    def __init__(self, name: str, defend_type: str, default_value: int, coef: Dict[str, int], duration: int, cooldown: int):
        self.name = name
        self.type = defend_type
        self.default_value = default_value
        self.coef = coef
        self.duration = duration
        self.cooldown = cooldown
        self.cur_cooldown = 0

    def do_defend(self, stats: Dict[str, Any]) -> Dict[str, Any]:
        total_value = self.default_value

        for key, value in self.coef.items():
            total_value += stats.get(key, 0) * value

        self.cur_cooldown += self.cooldown

        return {
            "type": self.type,
            "value": total_value,
        }

    def is_available(self) -> bool:
        return self.cur_cooldown == 0

    def reduce_cooldown(self, value: int) -> None:
        self.cur_cooldown = max(self.cur_cooldown - value, 0)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "type": self.type,
            "defaultValue": self.default_value,
            "coef": self.coef,
            "duration": self.duration,
            "cooldown": self.cooldown,
            "curCooldown": self.cur_cooldown
        }
